cryptography/src/bls12381/dkg.rs

```diff
@@ -1305,44 +1256,37 @@ fn select<V: Variant, P: PublicKey>(
 
 struct ObserveInner<V: Variant, P: PublicKey> {
     output: Output<V, P>,
-    weights: Option<BTreeMap<u32, Weight>>,
+    weights: Option<Interpolator<P, Scalar>>,
 }
 
 impl<V: Variant, P: PublicKey> ObserveInner<V, P> {
     fn reckon(
         info: Info<V, P>,
-        selected: Vec<(P, DealerLog<V, P>)>,
+        selected: Map<P, DealerLog<V, P>>,
         concurrency: usize,
     ) -> Result<Self, Error> {
         let (public, weights) = if let Some(previous) = info.previous.as_ref() {
-            let (indices, commitments) = selected
+            let weights = previous
+                .public()
+                .mode()
+                .subset_interpolator(previous.players(), selected.keys())
+                .expect("the result of select should produce a valid subset");
+            let commitments = selected
                 .into_iter()
-                .map(|(dealer, log)| {
-                    let index = previous
-                        .players()
-                        .index(&dealer)
-                        .expect("select checks that dealer exists, via our signature");
-                    (index, (index, log.pub_msg.commitment))
-                })
-                .collect::<(Vec<_>, BTreeMap<_, _>)>();
-
-            let weights =
-                poly::compute_weights(info.mode, NZU32!(previous.players.len() as u32), indices)
-                    .expect("should be able to compute weights");
-            let public = recover_public_with_weights::<V>(
-                &commitments,
-                &weights,
-                info.threshold(),
-                concurrency,
-            );
+                .map(|(dealer, log)| (dealer, log.pub_msg.commitment))
+                .try_collect::<Map<_, _>>()
+                .expect("Map should have unique keys");
+            let public = weights
+                .interpolate(&commitments, concurrency)
+                .expect("select checks that enough points have been provided");
             if previous.public().public() != public.constant() {
                 return Err(Error::DkgFailed);
             }
             (public, Some(weights))
         } else {
             let mut public = Poly::zero();
-            for (_, log) in selected.iter() {
-                public.add(&log.pub_msg.commitment);
+            for log in selected.values() {
+                public += &log.pub_msg.commitment;
             }
             (public, None)
         };
```

Comment:
Can't comment on it directly but `info.mode` should be used in `Output`?

Reply:
Yep that was a mistake